docs/Lifecycling.md

```diff
@@ -0,0 +1,148 @@
+
+In the example above, `t_1` would be a fixing date and `t_2` a payment date, as well as the expiry date.
+
+### Determining event dates
+
+Once we have acquired a contract `c` at a time `t_0`, the next event date is due to either
+
+- an election being made on an `Anytime` node by the contract holder
+
+- a condition within a `When` or `Until` node becoming `True`
+
+- the contract becoming equivalent to the `Zero` contract (e.g. if the probability of the condition within a `When` node becoming `True` is zero)
+
+Event dates related to `When` and `Until` conditions are modelled as stopping times of stochastic processes: these processes need to be continuously observed in order to determine when the event occurs. An exception to this are cases where the next event date is a deterministic time, for instance when using the condition `t ≥ t_1`: in this case, even before `t_1` we know that `t_1` will be an event date.
+
+### Upper bound for the expiry date
+
+TODO: describe cases when an upper bound to the expiry date of the contract is deterministically known.
+
+### Desirable features of a lifecycle functionality
+
+The lifecycle function takes as an input
+
+- a claim `c`
+
+- a date `t`
+
+- potentially, an `election` made by the contract holder
+
+and returns
+
+- the evolved claim
+
+- the amounts to be settled
+
+Some features for this function are desirable:
+
+1. if `t` does not correspond to an event date, the contract should not mutate
+
+2. to reduce the number of contracts that need to be created, event dates that are exclusively fixing dates should not mutate the contract.
+
+The second feature facilitates handling of contracts with many fixing dates but only few payment dates (e.g. barrier options, OIS swaps). However, it adds additional complexity: when we lifecycle at a payment or election date, we need to be able to collect the cumulated effects of all previous fixing events.
+
+### Implementation assumptions
+
+The features described above are difficult to implement in practice. The following assumptions are made to simplify the implementation.
+
+#### Boolean Conditions
+
+The boolean conditions within a `When` or `Until` node can be of the form
+
+- `t ≥ t_0`
+
+- `o1(t) ≤ o2(t)` for two observables `o1`, `o2`
+
+The former allows for the deterministic determination of the first instant a condition is met. The latter might be of stochastic nature.
+
+#### Contract acquisition date
+
+Assume that we invoke the lifecycle function on a contract `c1` at a time `t`:
+
+- if `c1` is a `When {t ≥ t_0} c2` with `t_0` ≤ `t1`, then `c2` is acquired at time `t_0`
+
+- if `c1` is a `When {o1 ≤ o2} c2` with `t_0` ≤ `t1` and the condition is `True` at time `t`, then `c2`is acquired at time `t`
+
+This means that, in the case of a stochastic condition, we assume `t` to be the first instant when the condition has become `True`.
```

Comment:
Is it worth elaborating on why we have three separate lifecycle functions?

You mentioned above:

> stochastic processes: these processes need to be continuously observed

The practicalities of this mean that we can't guarantee continuous observation, and hence we need to be able to lifecycle stuff a posteriori.

Also consider what happens if for operational reasons we're unable to lifecycle an `Or` condition in time. Then we have a liveness issue. This might happen in other cases too (e.g. missing market data).

Reply:
agree, it must always be possible to lifecycle "late" (a posteriori). To lifecycle an `Or` node "late", we would need to make sure (by other means) that the actual decision (being fed into the `Or`) was done no later than the acquisition time of the `Or`. Should also work similarly for `Anytime`.

Reply:
I will elaborate on why we have three separate lifecycle functions, I think it makes a lot of sense. 

That said, I have the feeling we do not really need `expire` and think that `lifecycle` and `exercise` are sufficient. More details of what I mean in https://github.com/digital-asset/contingent-claims/issues/69